cli-plugins/plugin/plugin.go

```diff
@@ -24,12 +32,45 @@ import (
 // called.
 var PersistentPreRunE func(*cobra.Command, []string) error
 
+// closeOnParentSocketClose connects to the socket specified
+// by the CLI_PLUGIN_EXIT_SOCKET env var, if present, and attempts
+// to read from it until it receives an EOF, which signals that
+// the CLI is going to exit and the plugin should also exit.
+func closeOnParentSocketClose() {
+	socketAddr, ok := os.LookupEnv(CliPluginCloseSocketEnvKey)
+	if !ok {
+		// if a plugin compiled against a more recent version of docker/cli
+		// is executed by an older CLI binary, ignore missing environment
+		// variable and behave as usual
+		return
+	}
+	addr, err := net.ResolveUnixAddr("unix", socketAddr)
+	if err != nil {
+		return
+	}
+	cliCloseConn, err := net.DialUnix("unix", nil, addr)
+	if err != nil {
+		return
+	}
+
+	go func() {
+		for {
+			b := make([]byte, 1)
+			_, err := cliCloseConn.Read(b)
+			if errors.Is(err, io.EOF) {
+				os.Exit(0)
```

Comment:
`os.Exit` is very abrupt; no goroutine stacks are unwound and no deferred calls are invoked. Should the plugin [be given a chance to clean up and exit gracefully?](https://pkg.go.dev/github.com/spf13/cobra#Command.ExecuteContext) Any plugin that is fine with ungraceful termination [could easily opt in](https://pkg.go.dev/context#AfterFunc):
```go
context.AfterFunc(cmd.Context(), func() { os.Exit(0) })
```

Reply:
That's a fair point – however, making the new behavior opt-in sounds like a straightforward way to have all plugins ignore it and keep misbehaving. How does opt-out sound? We could change the API for `plugin.Run` and force plugins to pass a "shutdown" function, that will be called when the CLI is going to exit, and have that be `os.Exit(0)` if the plugin passes in `nil`. Or we can have `plugin.Run` take a channel, or some other notification mechanism, as long as that involves an API change that forces plugins to explicitly opt-out/overwrite shutdown behavior, and not just ignore it.

Reply:
@thaJeztah thoughts? (re: opt-in vs opt-out, and how the API could look)

Reply:
I don't think it would be worth changing the plugin API just to guard against plugins that do not respect the graceful termination signal. A plugin could opt-in to graceful shutdown and hang anyway, e.g. because of a deadlock. If plugins misbehave then the plugins should be fixed. Aren't they all first-party plugins anyway?

Consider Option C: give all plugins a chance to exit upon context cancelation, with a deadline. That would cover both the "plugin ignores cancelation" and "plugin cancelation handling is bugged" cases, without any API changes.
```go
go func() {
    time.Sleep(gracefulExitGracePeriod) // e.g. 15s
    logrus.Fatal("plugin did not gracefully terminate in time")
}()
```

Reply:
Why is changing the API a negative, though? Compared to "Option C", changing the API will necessarily make plugin authors aware of the change in behavior and that they should now handle shutdown behavior (by being forced to pass in a shutdown function, for example), whereas with "Option C" all plugins will happily still compile, but since they won't necessarily be handling the case where their `CommandContext` gets cancelled, they won't shutdown and will cause the
```go
go func() {
    time.Sleep(gracefulExitGracePeriod) // e.g. 15s
    logrus.Fatal("plugin did not gracefully terminate in time")
}()
```
block to execute and print the message, at which point they'll have to dig in to the changes, find this PR and that they now have to handle the case where their `CommandContext` gets cancelled, and then make the appropriate changes.

I think communicating contract changes via API changes is a lot clearer, but there are probably some downsides to that change that I'm missing (please let me know 😅).

Reply:
> We could change the API for plugin.Run and force plugins to pass a "shutdown" function, that will be called when the CLI is going to exit, and have that be os.Exit(0) if the plugin passes in nil. Or we can have plugin.Run take a channel, or some other notification mechanism, as long as that involves an API change that forces plugins to explicitly opt-out/overwrite shutdown behavior, and not just ignore it.

Not sure if a "shutdown" function at the top level would be convenient to use. Looking at what [buildx](https://github.com/docker/buildx/blob/5b5c4c8c9df55e133c39cce8153e0fa9fc6f60c4/commands/create.go#L50) and [compose](https://github.com/docker/compose/blob/30e80d244003761d5f1563f7f6dcd8aab515de0a/cmd/compose/compose.go#L74) - they create their own context that gets cancelled when the plugins receives the termination signal. So to handle the graceful exit they would need to introduce additional glue between the that shutdown function and their custom context.

Having the `CommandContext` being cancelled would be much more convenient way for the consumers of this API to handle the graceful exit correctly - assuming they pass the `CommandContext` and use the `context` idiomatically, they probably wouldn't even need to modify their code at all in most cases.
It will stop every cli plugin to invent its own shutdown mechanism and would provide a more unified behavior.

I wouldn't drop the `os.Exit` completely though - I think we still want to do it in case the plugin ignores the context cancellation and doesn't stop in a specified time (10s?).

I think that's what Cory described as "Option C" 😄. IMO that's the best solution, not because it doesn't break the API (I think it's fine to have a breaking changes in that API, specifically because they're mostly first party), but because it's simply easier and more idiomatic for the consumers.

Reply:
I couldn't have said it better myself @vvoland. Also, #4700 is very timely and relevant. Contexts need to be plumbed through everywhere for tracing, regardless. If a plugin has been missed, a failure to gracefully exit is a much more visible sign than otel traces being less detailed than expected that contexts have not been plumbed through properly.

Reply:
> I think that's what Cory described as "Option C" 😄. IMO that's the best solution, not because it doesn't break the API (I think it's fine to have a breaking changes in that API, specifically because they're mostly first party), but because it's simply easier and more idiomatic for the consumers.

Gotcha, this sounds good to me. Thanks all for the discussion